src/opnsense/mvc/app/controllers/OPNsense/CaptivePortal/Api/MasqueradeController.php

```diff
@@ -0,0 +1,174 @@
+     * @return array
+     */
+    private function clientSession($zoneid, $clientIp = null)
+    {
+        // If we've not been passed a client IP then work it out
+        if($clientIp == null)
+        {
+            $clientIp = $this->getClientIp();
+        }
+        
+        $backend = new Backend();
+        $allClientsRaw = $backend->configdpRun(
+            "captiveportal list_clients",
+            array($zoneid, 'json')
+        );
+        $allClients = json_decode($allClientsRaw, true);
+        if ($allClients != null) {
+            // search for client by ip address
+            foreach ($allClients as $connectedClient) {
+                if ($connectedClient['ipAddress'] == $clientIp) {
+                    // client is authorized in this zone according to our administration
+                    $connectedClient['clientState'] = 'AUTHORIZED';
+                    return $connectedClient;
+                }
+            }
+        }
+
+        // return Unauthorized including authentication requirements
+        $result = array('clientState' => "NOT_AUTHORIZED", "ipAddress" => $clientIp);
+        $mdlCP = new CaptivePortal();
+        $cpZone = $mdlCP->getByZoneID($zoneid);
+        if ($cpZone != null && trim((string)$cpZone->authservers) == "") {
+            // no authentication needed, logon without username/password
+            $result['authType'] = 'none';
+        } else {
+            $result['authType'] = 'normal';
+        }
+        return $result;
+    }
+
+    /**
+     * determine client's ip address
+     */
+    private function getClientIp()
+    {
+        // determine orginal sender of this request
+        $trusted_proxy = array(); // optional, not implemented
+        if ($this->request->getHeader('X-Forwarded-For') != "" &&
+            (
+            explode('.', $this->request->getClientAddress())[0] == '127' ||
+            in_array($this->request->getClientAddress(), $trusted_proxy)
+            )
+        ) {
+            // use X-Forwarded-For header to determine real client
+            return $this->request->getHeader('X-Forwarded-For');
+        } else {
+            // client accesses the Api directly
+            return $this->request->getClientAddress();
+        }
+    }
+
+    /**
+     * logon client to zone, must use post type of request
+     * @param int|string zone id number
+     * @return array
+     */
+    public function logonAction($zoneid = 0)
+    {
+        // get username, IP and auth server from post
+        $userName = $this->request->getPost("user", "striptags", null);
+        $clientIp = $this->request->getPost("ip", "striptags", null);
+        $authServer = $this->request->getPost("server", "striptags", null);
+        $timeout = $this->request->getPost("timeout", "striptags", null);
+        
+        // If an explicit client IP wasn't provided
+        if(!$clientIp)
+        {
+            $clientIp = $this->getClientIp();
```

Comment:
If you want to allow a client then this line is wrong - you should return an error instead.